### PR TITLE
Migrated all components to native classes. Requires Babel-Polyfill for IE11

### PR DESCRIPTION
**Pull request by simonihmig**
Migrate all classes to native ES6 classes. 

Assuming all tests pass, this should be a backwards compatible change for "normal" usage. When users have extended from e-b classes, this might be a breaking change though, due to subtle semantic differences. 

Especially the use of class fields for applying default values introduces noticeable changes, as the defaults are effectively applied in the constructor to the class *instance*, while in classic classes they were part of the prototype. This will make overriding props more difficult, as the field initializer will override values that live on the prototype (e.g. a classic-style subclass). Especially when the overridden property is a getter-only (i.e. readonly) property/CP, hence the experiment with the `@defaulValue` decorator (which will create a getter, so the default is again living on the prototype).

To Do:
- [x] figure out if we really should keep `@classic`
- [x] figure out if we need more instances of `@defaultValue` or some other way to apply defaults
- [ ] test with real apps

### CODE REVIEW

**Comment by simonihmig** [comment]
@jelhan found a way to run the codemod on the addon's special structure. And after fixing a few things, things seem to work fine - at least locally, let's see what Travis thinks...

Given the fact that extending from those classes might break in some cases (see comments above), I am thinking if we should hold back 3.0 a bit to get this to land before!?

File: addon/templates/components/common/bs-form.hbs [hunk 1]
```diff
@@ -7,14 +7,14 @@
       showAllValidations=this.showAllValidations
       disabled=this.disabled
       readonly=this.readonly
-      onChange=(action "change")
+      onChange=(action "elementChanged")
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply addon/templates/components/common/bs-form.hbs]: Overlooked. That wasn't done by the codemod, so has to be done manually. Will fix it, good catch!

**Comment by simonihmig** [comment]
> Is this only to support people extending this one the old way using .extend({})? 

No, to the opposite. It was required in some cases when extending natively. And that in our own codebase: the `BsButtongroup` component yields an extended `BsButton`, and overrides the `active` property (usually defaulting to `false`) with its own button-group related logic, implemented now as a native class getter-only CP: https://github.com/kaliber5/ember-bootstrap/pull/888/files#diff-019ceff946d1b8dd515d3923cca59f0fR36

But when the (parent) `BsButton#active` initializer tries to set the default value to `false`, it does that on the instance, thus trying to set the above getter-only (i.e. readonly), which results in: 

image

The added `@defaultValue` solution/workaround/hack (?) prevents that, by making the default value live on the prototype as a getter, similar to how it was before with classic classes, so the `active` CP of the subclass correctly overrides (shadows) the default value.

> I wouldn't consider extending Ember Bootstrap classes as part of public API. 

Would disagree here. Sometimes it is required for customization. And the validation-supporting "addon-addons" like `ember-bootstrap-cp-validations` have to do this as well: https://github.com/offirgolan/ember-bootstrap-cp-validations/blob/master/addon/components/bs-form.js#L6

> Supporting extending through native classes only would be fine for me.

Not the issue, see above.

> I'm not sure about the naming of sub-components classes. Having generic names like Title or Body may make debuging more difficult. Maybe AccordionItemTitle and AccordionItemBody would be more helpful if something goes wrong?

Agree, was done by the codemod automatically. Basically should match what the `@class` YUIDoc annotation writes I think.

File: addon/components/base/bs-dropdown/button.js [hunk 2]
```diff
@@ -15,7 +15,7 @@ import { schedule } from '@ember/runloop';
 @classNames('dropdown-toggle')
 export default class DropdownButton extends Button {
   didReceiveAttrs() {
-    super.didReceiveAttrs(...arguments);
+    // super.didReceiveAttrs(...arguments);
```
Comment (jelhan) [inline addon/components/base/bs-dropdown/button.js]: `didReceiveAttrs()` is missing a noop implementation on `@ember/component`, isn't it?

File: addon/components/base/bs-contextual-help.js [hunk 3]
```diff
@@ -65,14 +68,16 @@ export default class ContextualHelp extends Component {
...
   visible = false;
 
   /**
    * @property inDom
    * @type boolean
    * @private
    */
-  inDom = false;
+  @defaultValue
+  inDom = this.get('visible') && this.get('triggerTargetElement');
```
[the remark this answers is not in the document]
Comment (simonihmig) [reply addon/components/base/bs-contextual-help.js]: Yes, this does work. And it is set later, so a normal `and()` would not work (or probably cause the "clobbering a CP" deprecation)

File: addon/components/base/bs-dropdown/button.js (hunk 2 above)
Comment (simonihmig) [reply addon/components/base/bs-dropdown/button.js]: IIRC this was the problem, yes. `this._super` seems to always have a noop implementation.

**Comment by simonihmig** [comment]
It's all green now! 🎉 

There is one breaking change here, which is that IE11 requires the Babel Polyfill now, due to `ember-decorators` doing `for ... or` loops that requires `Symbol` support which IE11 doesn't have. But I think that's pretty much acceptable, given that a) many users probably don't have to support IE11 anymore, and b) those that do will probably have included the polyfill anyway.

@jelhan I think I will merge this, and cut another RC release, to test this with some real-world apps to make sure no additional unexpected issues come up caused by this bigger refactoring.

**Comment by jelhan** [comment]
> It's all green now! 

Awesome! :clap: 
 
> There is one breaking change here, which is that IE11 requires the Babel Polyfill now, due to `ember-decorators` doing `for ... or` loops that requires `Symbol` support which IE11 doesn't have.

Yeah, that should be fine if we document it well. I think adding this to _compatibility_ section in readme and [setup instructions](https://www.ember-bootstrap.com/#/getting-started/setup) in docs should be fine. And mentioning in release notes as a breaking change of course.

> a) many users probably don't have to support IE11 anymore

I fear you are to optimistic.

> @jelhan I think I will merge this, and cut another RC release, to test this with some real-world apps to make sure no additional unexpected issues come up caused by this bigger refactoring.

:+1:

**Pull request merged**